### PR TITLE
Bugfix/groupby datetime issue

### LINKED ISSUE #28247: [Bug][Regression] df.groupby.apply fails under specific conditions
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
import datetime

def get_vals(x):
    return pd.Series([0,1,2], index=[2000, 2001, 2002])

b = list(range(0,3))*2
y = list(range(2000,2003))*2
df = pd.DataFrame({'b':b,'y':y}) 
df['date'] = pd.to_datetime(df['y'].apply(lambda x: datetime.date(x, 1, 1)))


print(pd.__version__)

print(df)

df.groupby(['b']).apply(lambda x: get_vals(x))

```
#### Problem description

The above code gives an error (KeyError: 0) in pandas 0.25.1, while it runs as expected in pandas 0.24.2

The apply function returns a timeseries.
I found that triggering the error is related to having a datetime column in the dataframe. Without that column it does not throw an error. 

#### Expected Output
No error and the following output, as can be verified with pandas version 0.24.2:
image

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
Environment of 0.25.1 that I tested on:

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.4.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 0.25.1
numpy            : 1.16.4
pytz             : 2019.2
dateutil         : 2.8.0
pip              : 19.2.2
setuptools       : 41.0.1
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : 2.7.6.1 (dt dec pq3 ext lo64)
jinja2           : 2.10.1
IPython          : 7.7.0
pandas_datareader: None
bs4              : None
bottleneck       : 1.2.1
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.3.1
sqlalchemy       : 1.3.7
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

Environment of 0.24.2 that I tested on:
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.3.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 10, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.24.2
pytest: 5.1.1
pip: 19.2.3
setuptools: 41.2.0
Cython: 0.29.13
numpy: 1.16.4
scipy: 1.3.1
pyarrow: None
xarray: None
IPython: 7.7.0
sphinx: 2.2.0
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2019.2
blosc: None
bottleneck: 1.2.1
tables: 3.5.2
numexpr: 2.7.0
feather: None
matplotlib: 3.1.1
openpyxl: 2.6.3
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.2.0
lxml.etree: 4.4.1
bs4: 4.8.0
html5lib: 1.0.1
sqlalchemy: 1.3.7
pymysql: None
psycopg2: None
jinja2: 2.10.1
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.7.0
gcsfs: None
</details>

### PR DESCRIPTION
**Pull request by sidharthann**
- [x] closes #28247
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: doc/source/whatsnew/v1.0.0.rst [hunk 1]
```diff
@@ -181,6 +181,7 @@ Datetimelike
 - Bug in :class:`Series` and :class:`DataFrame` with integer dtype failing to raise ``TypeError`` when adding or subtracting a ``np.datetime64`` object (:issue:`28080`)
 - Bug in :class:`Week` with ``weekday`` incorrectly raising ``AttributeError`` instead of ``TypeError`` when adding or subtracting an invalid type (:issue:`28530`)
 - Bug in :class:`DataFrame` arithmetic operations when operating with a :class:`Series` with dtype `'timedelta64[ns]'` (:issue:`28049`)
+- Bug in :func:`pandas.core.groupby.generic._recast_datetimelike_result` integer indexes for columns referenced by name (:issue:`28247`)
```
Comment (gfyoung) [inline doc/source/whatsnew/v1.0.0.rst]: Let's discuss how a publicly-facing function (i.e. one that isn't prefixed by an underscore) was affected by this bug instead in the entry.

Comment (sidharthann) [reply doc/source/whatsnew/v1.0.0.rst]: Thanks, I've updated.

File: doc/source/whatsnew/v1.0.0.rst [hunk 2]
```diff
@@ -181,7 +181,7 @@ Datetimelike
 - Bug in :class:`Series` and :class:`DataFrame` with integer dtype failing to raise ``TypeError`` when adding or subtracting a ``np.datetime64`` object (:issue:`28080`)
 - Bug in :class:`Week` with ``weekday`` incorrectly raising ``AttributeError`` instead of ``TypeError`` when adding or subtracting an invalid type (:issue:`28530`)
 - Bug in :class:`DataFrame` arithmetic operations when operating with a :class:`Series` with dtype `'timedelta64[ns]'` (:issue:`28049`)
--
+- Bug in :func:`pandas.core.groupby.generic._recast_datetimelike_result` causing ``KeyError`` in :func:`pandas.core.groupby.generic.SeriesGroupBy.apply` when a column index in the original DataFrame is datetime like(:issue:`28247`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.0.0.rst]: no space between "datetime" and "like", space after "like".

The only test is for a datetime object.  Does this also work for the other things we usually call "datetimelike"?

Comment (sidharthann) [reply doc/source/whatsnew/v1.0.0.rst]: hi, i have updated as per your suggestion, and added a few more test cases

File: pandas/tests/groupby/test_apply.py [hunk 3]
```diff
@@ -657,3 +657,31 @@ def test_apply_with_mixed_types():
 
     result = g.apply(lambda x: x / x.sum())
     tm.assert_frame_equal(result, expected)
+
+
+def test_apply_datetime_issue():
+    # GH-28247
```
Comment (jreback) [inline pandas/tests/groupby/test_apply.py]: can you add a 1-line comment on what the issue is

File: doc/source/whatsnew/v1.0.0.rst [hunk 4]
```diff
@@ -182,7 +182,7 @@ Datetimelike
 - Bug in :class:`Series` and :class:`DataFrame` with integer dtype failing to raise ``TypeError`` when adding or subtracting a ``np.datetime64`` object (:issue:`28080`)
 - Bug in :class:`Week` with ``weekday`` incorrectly raising ``AttributeError`` instead of ``TypeError`` when adding or subtracting an invalid type (:issue:`28530`)
 - Bug in :class:`DataFrame` arithmetic operations when operating with a :class:`Series` with dtype `'timedelta64[ns]'` (:issue:`28049`)
--
+- Bug in :func:`pandas.core.groupby.generic._recast_datetimelike_result` causing ``KeyError`` in :func:`pandas.core.groupby.generic.SeriesGroupBy.apply` when a column index in the original DataFrame is datetimelike (:issue:`28247`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.0.0.rst]: can you not reference internal functions; and make this what a user cares about

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/groupby/test_apply.py [hunk 5]
```diff
@@ -657,3 +657,34 @@ def test_apply_with_mixed_types():
...
+    # groupby-apply throws an error if one of the columns in the DataFrame
+    #   is a datetime object and the column labels are different from
+    #   standard int values in range(len(num_columns))
+
+    df = pd.DataFrame({"a": ["foo"], "b": [datetime.today()]})
```
Comment (jreback) [inline pandas/tests/groupby/test_apply.py]: can you parameterize these tests, only thing changing is the b column which you can pass in the paramters

**Review by jreback — CHANGES_REQUESTED** [review]
change in the tests, ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @sidharthann

**Pull request merged**